paddle/fluid/platform/device_tracer.cc

```diff
@@ -195,9 +232,15 @@ class DeviceTracerImpl : public DeviceTracer {
  public:
   DeviceTracerImpl() : enabled_(false) {}
 
-  void AddAnnotation(uint64_t id, const std::string &anno) {
-    std::lock_guard<std::mutex> l(trace_mu_);
-    correlations_[id] = anno;
+  void AddAnnotation(uint64_t id, Event *event) {
+    static thread_local std::list<std::pair<uint64_t, Event *>>
+        *local_correlations_pairs = nullptr;
+    if (local_correlations_pairs == nullptr) {
+      std::lock_guard<std::mutex> l(trace_mu_);
+      correlations_pairs.emplace_back();
+      local_correlations_pairs = &correlations_pairs.back();
+    }
```

Comment:
line 238~242 is not thread-safe.

Reply:
I see, thanks.
Another comment is all variables declared with the thread_local keyword have thread storage duration, so `static` is unnecessary in line236.

Reply:
Thanks, fixed